Relax `Allocator` bounds into pin-safe trait

djkoloski opened the pull request:
Because `Allocator` requires that returned memory blocks remain valid
until the instance and all clones are dropped, some types of allocators
cannot fulfill the safety requirements for the trait. This change
relaxes the safety requirements of `Allocator` to blocks remaining valid
until the allocator becomes unreachable and moves the bound into a
separate `PinSafeAllocator` trait.

It also relaxes some overly cautious bounds on the `Unpin` impls for
`Box` since unpinning a box doesn't require any special consideration.

djkoloski committed:
Relax `Allocator` bounds into pin-safe trait

Because `Allocator` requires that returned memory blocks remain valid
until the instance and all clones are dropped, some types of allocators
cannot fulfill the safety requirements for the trait. This change
relaxes the safety requirements of `Allocator` to blocks remaining valid
until the allocator becomes unreachable and moves the bound into a
separate `PinSafeAllocator` trait.

It also relaxes some overly cautious bounds on the `Unpin` impls for
`Box` since unpinning a box doesn't require any special consideration.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5243282437?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/875742b182b4e08f3c1f77f054ba8e7e0bc7b6ba/checks/5243282437/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.15s
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
 Documenting core v0.0.0 (/checkout/library/core)
error: unresolved link to `::alloc::rc::Rc`
    |
    |
412 | /// `Allocator` in shared pointer like [`Rc`](::alloc::rc::Rc) or [`Arc`](::alloc::sync::Arc).
    |                                               ^^^^^^^^^^^^^^^ no item named `` in scope
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`

error: unresolved link to `::alloc::sync::Arc`
    |
    |
412 | /// `Allocator` in shared pointer like [`Rc`](::alloc::rc::Rc) or [`Arc`](::alloc::sync::Arc).
    |                                                                           ^^^^^^^^^^^^^^^^^^ no item named `` in scope
error: could not document `core`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name core library/core/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc --error-format=json --json=diagnostic-rendered-ansi,future-incompat --markdown-css rust.css --markdown-no-toc -Z unstable-options --resource-suffix 1.60.0 --index-page /checkout/src/doc/index.md -C metadata=4599449636879869 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.60.0-nightly
  (6d242799a
  2022-02-18)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")'` (exit status: 1)
```

</details>

CAD97 commented on library/core/src/alloc/mod.rs:
```diff
@@ -87,7 +87,7 @@ impl fmt::Display for AllocError {
 ///
 /// * Memory blocks returned from an allocator must point to valid memory and retain their validity
-///   until the instance and all of its clones are dropped,
+///   until the instance and all of its clones become unreachable,
```
This is, of course, the hard part of this PR.

I'm thinking of the guarantee in a similar way to that of references, in that it's a "while the handle exists." Of course, under the Stacked Borrows model (and, to an extent, NLL), it's _not_ when the borrow exists, but until when the borrow is last used.

So there're two ways I currently see to explain the weaker (not pin safe) guarantee, borrowing language from references:

- The memory regions are valid for the lifetime of the allocator. This is complicated by the written `'static` lifetime not actually being "the lifetime of the allocator" in this fashion, but an upper bound, so probably not a good way to explain it.
- The memory regions are valid until the point when the last clone is last used.

Effectively, what we're looking for is the dual of the pinning guarantee.

So I _think_ how I'd draft this is

> Memory blocks must retain their validity until the allocator instance and all of its clones are dropped, forgotten, leaked, or otherwise never accessed again. (In other words, when a lifetime held by the allocator instance is allowed to expire.)

(And now writing that I'm interested in how the `#[may_dangle]` borrowck eyepatch interacts with the guarantee, since that (very roughly) allows a lifetime to expire before the container. IIUC it doesn't, since the eyepatch refers to the lifetime within a generic argument, but I'm never certain w.r.t. may_dangle.)

CAD97 reviewed (commented):
I believe `Box::leak` also needs to be bound by `A: PinSafeAllocator`. Otherwise, this matches what I was looking for with extracting the pin safe guarantee into its own extension trait.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5248894582?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/32fd054a751f5414539f19ff1b6c168f4d5efb59/checks/5248894582/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling alloc v0.0.0 (/checkout/library/alloc)
   Compiling cfg-if v0.1.10
   Compiling adler v0.2.3
   Compiling rustc-demangle v0.1.21
error[E0277]: the trait bound `A: core::alloc::PinSafeAllocator` is not satisfied
     |
     |
1094 |         (Unique::from(Box::leak(b)), alloc)
     |                       --------- ^ expected an implementor of trait `core::alloc::PinSafeAllocator`
     |                       required by a bound introduced by this call
     |
     |
note: required by a bound in `Box::<T, A>::leak`
     |
     |
1147 |     pub const fn leak(b: Self) -> &'static mut T
     |                  ---- required by a bound in this
1148 |     where
1149 |         A: PinSafeAllocator + ~const PinSafeAllocator,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `Box::<T, A>::leak`
     |
     |
1094 |         (Unique::from(Box::leak(&b)), alloc)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

djkoloski commented:
Looks like changing `leak` comes with some baggage. I'll have to take a closer look later on.

djkoloski commented:
I pushed through some changes to `Box::leak`. There were some comments about MIRI in there, so it might be a good idea to get these tested properly under it.

compiler-errors commented on library/alloc/src/boxed.rs:
```diff
@@ -543,7 +543,7 @@ impl<T, A: Allocator> Box<T, A> {
     pub const fn pin_in(x: T, alloc: A) -> Pin<Self>
     where
-        A: 'static + ~const Allocator + ~const Drop,
+        A: PinSafeAllocator + ~const Allocator + ~const PinSafeAllocator + ~const Drop,
```
I don't think you need both of these bounds, just the `~const PinSafeAllocator` one. `A: ~const PinSafeAllocator` implies `A: PinSafeAllocator`.

djkoloski replied:
Ah, you're right. I'm new to the `~const Trait` syntax. This is fixed now.

compiler-errors commented on library/alloc/src/boxed.rs:
```diff
@@ -1146,9 +1150,13 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
     pub const fn leak<'a>(b: Self) -> &'a mut T
     where
-        A: 'a,
+        A: PinSafeAllocator + ~const PinSafeAllocator,
```
Do you mind changing this to `A: ~const PinSafeAllocator` everywhere else too?

djkoloski replied:
Fixed

bors commented:
:umbrella: The latest upstream changes (presumably #94901) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

djkoloski commented:
@rustbot label +T-libs-api

tmandry commented:
cc @rust-lang/wg-allocators

djkoloski commented:
A summary of this PR's history and motivation:

## History

In #79327, many `Pin` APIs in `Box` were restricted to `A: 'static` based on this example:

```rust
let alloc = MyAlloc(/* ... */);
let pinned = Box::pin_in(42, alloc);
mem::forget(pinned); // Now `value` must live forever
// Otherwise `Pin`'s invariants are violated, storage invalidated
// before Drop was called.
// borrow of `memory` can end here, there is no value keeping it.
drop(alloc); // Oh, value doesn't live forever.
```

In #90822, I proposed reverting these restrictions with the following reasoning:

> Forgetting the box will also forget the allocator, which will prevent the memory from being freed and therefore reused. If you move a clone of `alloc` into the box, you're still covered by the safety conditions for `Allocator`:
>
> > - Memory blocks returned from an allocator must point to valid memory and retain their validity until the instance and all of its clones are dropped,
>
> At least one clone of the allocator must be forgotten along with the pinned box. This is also why `Box::leak` is sound with a restriction of `A: 'a`; the allocator is not dropped when the box is leaked.

To summarize, `alloc` was moved into `pinned` by `pin_in` and later forgotten with `mem::forget`. If forgetting `alloc` causes memory allocated by it to be invalidated, then it violates the safety requirements of `Allocator`.

I also noted:

> It is important to note that naive implementations of non-`'static` allocators cannot fulfill the safety requirements for `Allocator`. The allocator can be forgotten and its allocated memory will still be released at the end of its lifetime, which may cause the memory of undropped pinned values to be reused. The unsoundness is actually in the unsafe `Allocator` impl though, and not the boxed `Pin` APIs.

Which implies that the existing `unsafe impl<A: Allocator> Allocator for &A` is unsound. #94069 is a conservative PR that closes this soundness hole.

## Motivation

This PR is based on [previous discussions](https://github.com/rust-lang/rust/pull/94114#issuecomment-1043918590) regarding the `Allocator` API. It was observed that allocators are not allowed to invalidate memory when forgotten because pinned values are allowed to be created in that memory. This pin safety requirement could be separated into a new trait that augments `Allocator`, allowing `Allocator` to be relaxed for more use cases.

Specifically, this change allows `Allocator`s to invalidate allocated memory when they are forgotten or otherwise rendered inaccessible. This allows for a few new applications without compromising pin safety:

1. **Abort-free stack allocators:** Because of the safety requirements of `Allocator`, the only safe way to allocate memory from the stack is to shadow the memory and abort on drop if there are any outstanding allocations. After this change it is safe for `Allocator`s to return stack memory, it just cannot be pinned.
2. **Non-`'static` allocators:** All allocators must currently be `'static` and either be global or implement some kind of refcounting mechanism to prevent allocated memory from being invalidated until dropped. While these are still the only kinds of allocators that are pin-safe, non-`'static` allocators are still broadly useful.
3. **Composable pin-safety:** Allocators that are `'static` but not pin-safe can be augmented in several ways to make them pin-safe. Some of these ways include boxing them with `Box::new(alloc)`, reference counting them with `Rc::new(alloc)`, and extending their lifetimes with `Box::leak(Box::new(alloc))`. This allows a clean separation of pin-safety from allocation.

This change also adds clarity to the `Allocator` API, as many users tend to get confused by the subtle implications of the current safety requirements. Aside from the additional capabilities of a `PinSafeAllocator` trait, this new surface contrasts with `Allocator` and makes it easier to clearly articulate the safety implications of pinning.

Amanieu commented:
Another possibility would be to change the safety condition for `Allocator` to :
> Memory blocks returned from an allocator must point to valid memory and retain their validity until the instance and all of its clones are dropped, or for the lifetime of the allocator type, whichever is shorter.

(This could probably be worded better...)

This moves the ball back into `Box::pin_in`'s camp to required `A: 'static`, but it removes a footgun for implementers of non-`'static` allocators.

I'm hesitant to add more complexity to the `Allocator` API just to support pinning, which is a fairly niche operation.

djkoloski commented:
I think it would be sufficient to say:

> Memory blocks returned from an allocator must point to valid memory and retain their validity for the lifetime of the allocator type.

As a few examples:

- `'static` allocators must leak memory indefinitely (for a `'static` lifetime) if forgotten.
- Any allocator that manages backing memory with a non-static lifetime `'a` has its lifetime bounded by `'a`. Therefore, it's safe for the allocator to guarantee that the memory will not be repurposed during `'a`.

I've been prototyping out what this would look like in a crate and it seems like `'static` allocators would be compatible with `Pin`.

As an aside, I've also been trying out changing `Pin<P>` to only pin for the lifetime of `P` and that seems promising as well. I think there's a parallel here between generalizing the validity guarantees of `Allocator` and generalizing the pinning guarantees of `Pin`.

CAD97 commented:
Personally I think that extracting the pinning guarantees to a separate trait *simplifies* the main API w.r.t. pinning rather than complicates it for pinning, but if a simple single-trait solution works, then that's better.

Personally, I think I'd word it something along the lines of

> Memory blocks allocated by this allocator must remain valid for the lifetime of the allocator or until all clones of the allocator are dropped, whichever is sooner.
>
> In an equivalent formulation, given `A: 'a + Allocator`, `A` must allocate blocks compatible with the pinning guarantees of `Pin<Box<[u8], A>>`, which can be leaked into `Pin<&'a mut [u8]>`.

(This is sort of circular, as the purpose of this clarification is to make `Pin<Box<_, A>>` work, but `Box` is "how safe allocations work" and clearly communicates the `'a`-or-drop-whichever-is-first.)

-----

> As an aside, I've also been trying out changing `Pin<P>` to only pin for the lifetime of `P` and that seems promising as well.

This sounds sketchy, though, because how do you keep the moral equivalent of the following from blowing up? (Off topic for this thread, feel free to ping me on Zulip)

```rust
let mut future: impl Future;
'a: {
    Pin::new(&'a mut future).poll();
}
// 'a is over, future "isn't" pinned
// caveat emptor: yes it is, per current pin documentation
forget(future);
// reuse future's backing memory
// if the async runtime stored a pointer to the future, 💣
```

yaahc commented:
r? rust-lang/libs-api

JohnCSimon commented:
Ping from triage: 
@djkoloski 
Can you please address the merge conflict?

djkoloski committed:
Push new allocator trait through `BTreeMap`

The `BTreeMap` implementation was updated with support for custom
allocators, but many of the methods that take allocators do not require
that the provided allocator is the same allocator that is used for the
owning `BTreeMap`. This could lead to situations where nodes are created
with a different allocator, which would violate the safety conditions of
`deallocate` on drop. This was discovered because the changes to
`Box::leak` make it invalid to call without upgrading the allocator to
`A: PinSafeAllocator`.

This also updates `Box::pin_in` with the new `PinSafeAllocator` trait.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7489714824?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5cae951034acc536bbd48fccc2cbcf06cd99a165/checks/7489714824/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rand v0.7.3
    Checking core v0.0.0 (/checkout/library/core)
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking std v0.0.0 (/checkout/library/std)
error[E0133]: call to unsafe function is unsafe and requires unsafe block
   --> library/alloc/src/collections/btree/map/tests.rs:119:30
    |
119 |             self.root.insert(Root::new(*self.alloc)).bulk_push(iter, &mut self.length, *self.alloc);
    |                              ^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
    |
    = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function is unsafe and requires unsafe block
   --> library/alloc/src/collections/btree/map/tests.rs:119:13
    |
119 |             self.root.insert(Root::new(*self.alloc)).bulk_push(iter, &mut self.length, *self.alloc);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
    |
    = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function is unsafe and requires unsafe block
  --> library/alloc/src/collections/btree/node/tests.rs:71:21
   |
71 |     let mut root1 = NodeRef::new_leaf(Global);
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function is unsafe and requires unsafe block
  --> library/alloc/src/collections/btree/node/tests.rs:73:21
   |
73 |     let mut root1 = NodeRef::new_internal(root1.forget_type(), Global).forget_type();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function is unsafe and requires unsafe block
  --> library/alloc/src/collections/btree/node/tests.rs:74:17
   |
74 |     let root2 = Root::new(Global);
   |                 ^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior
For more information about this error, try `rustc --explain E0133`.
error: could not compile `alloc` due to 5 previous errors
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:01:27
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7489828637?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/f14ac694ea2d5792c09689ad3c270f593ed73afc/checks/7489828637/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---

######################################################################## 100.0%
extracting /checkout/obj/build/cache/2022-07-21/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/stage0
skip untracked path cpu-usage.csv during rustfmt invocations
Diff in /checkout/library/alloc/src/collections/btree/map/tests.rs at line 118:
         if !iter.is_empty() {
             // SAFETY: `self.alloc` is the allocator for this `BTreeMap`.
             unsafe {
-                self.root.insert(Root::new(*self.alloc))
-                    .bulk_push(iter, &mut self.length, *self.alloc);
+                self.root.insert(Root::new(*self.alloc)).bulk_push(
+                    iter,
+                    &mut self.length,
+                    *self.alloc,
             }
         }
     }
     }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/alloc/src/collections/btree/map/entry.rs" "/checkout/library/alloc/src/collections/btree/map/tests.rs" "/checkout/library/alloc/src/collections/btree/navigate.rs" "/checkout/library/alloc/src/collections/btree/borrow.rs" "/checkout/library/alloc/src/collections/btree/fix.rs" "/checkout/library/alloc/src/collections/btree/set_val.rs" "/checkout/library/alloc/src/collections/binary_heap.rs" "/checkout/library/alloc/src/collections/btree/search.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
thread '<unnamed>' panicked at 'tx.send(entry.into_path()) failed with sending on a closed channel', format.rs:166:17
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', /cargo/registry/src/github.com-1ecc6299db9ec823/ignore-0.4.18/src/walk.rs:1302:31
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', /cargo/registry/src/github.com-1ecc6299db9ec823/ignore-0.4.18/src/walk.rs:1302:31
thread '<unnamed>' panicked at 'tx.send(entry.into_path()) failed with sending on a closed channel', format.rs:166:17
```

</details>

djkoloski commented:
@JohnCSimon PR has been fixed up.

bors commented:
:umbrella: The latest upstream changes (presumably #100982) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

KittyBorgX commented:
@rustbot author
@djkoloski Hey! It looks like there are merge conflicts, so can you sort them out? A reviewer will review the code once you're done!

Dylan-DPC commented:
Closing this as inactive
